Stochastic Rounding Optimizers

crcrpar opened the pull request.

crcrpar committed:
Implement stochastic rounding features

crcrpar committed:
add PyTorch API & test

crcrpar committed:
fix order of arguments

crcrpar committed:
make tests redundant

crcrpar committed:
flake8

crcrpar committed:
add type stub file

crcrpar committed:
simplify stochastic rounding for tensor

mcarilli commented on aten/src/ATen/native/cuda/stochastic_rounding.cuh:
```diff
@@ -0,0 +1,67 @@
+// Ref: https://gitlab.com/riship11/stochastic-rounding
```
Remove all references to Apex and Nvidia internal repos.

crcrpar replied:
This is a public gitlab repository (we can access from anywhere) thus I thought it's allowed to refer.

crcrpar replied:
anyway remove this reference.

mcarilli commented on aten/src/ATen/native/cuda/stochastic_rounding.cuh:
```diff
@@ -0,0 +1,67 @@
+    return TWO_24;
+}
+
+// Natalia magic
```
Keep this comment.

ptrblck commented on torch/optim/sradam.py:
```diff
@@ -0,0 +1,91 @@
+
+# TODO(crcrpar): Decide whether to override `state_dict` method as
+# this optimizer tracks square root of `exp_avg_sq` and `max_exp_avg_sq`
+# that contradicts with its parent optimizer :class:`torch.optim.Adam`.
```
I think this would be a nice to have to make both optimizers compatible.

crcrpar replied:
Naively implemented the feature for the compatibility.

- [SRAdam](https://github.com/csarofeen/pytorch/pull/17/commits/5faf1a96caa3bf458d425a28cdfeecddf15537ba#diff-735fa5dfbcaa19fb4f36ace1f669c428)
- [SRAdamW](https://github.com/csarofeen/pytorch/pull/17/commits/5faf1a96caa3bf458d425a28cdfeecddf15537ba#diff-a5be7ac632ae7779a377f6ac40859580)
- [test](https://github.com/csarofeen/pytorch/pull/17/commits/5faf1a96caa3bf458d425a28cdfeecddf15537ba#diff-d237dd23d765aba7cb30b4c21a5e1a86)

ptrblck commented on test/test_optim.py:
```diff
@@ -1547,5 +1547,84 @@ def test_cosine_then_cyclic(self):
 
         self.assertLessEqual(last_lr, max_lr)
 
+class TestStochasticRoundingOptim(TestCase):
```
The tests might need to be skipped, if no CUDA device is available via `@unittest.skipIf(not torch.cuda.is_available(), 'No CUDA')`

ptrblck commented on test/test_stochastic_rounding.py:
```diff
@@ -0,0 +1,24 @@
+    delta_fp16 = math.pow(2, -10 + exponent if exponent >= -14 else -24)
+    threshold = 1e-6
+    diff = math.fabs(original_value - mean)
+    assert diff < threshold or diff < delta_fp16 / 2.0
```
Could you add some comments here, please (which value for denormals etc.)?

crcrpar replied:
Added note https://github.com/csarofeen/pytorch/pull/17/files#diff-02fa1f268871375277da2c690a676146R25-R28.

ptrblck commented on torch/optim/sradam.py:
```diff
@@ -0,0 +1,91 @@
+            with torch.no_grad():
+                loss = closure()
+
+        if grad_scaler is not None:
```
is `torch.cuda.current_device()` OK here? Could this create a CUDA context and allocate memory on GPU0, if we manually used GPU1 for all ops?

crcrpar replied:
So is it safer to get `get_inf` per parameter (or group) by `found_inf = device_found_inf_dict[param.device]`?

ptrblck replied:
Would it be possible to use the device of `scale` or another param?

mcarilli replied:
`GradScaler._check_inf_per_device` populates a different `found_inf` tensor on every device where the optimizer's parameters are.  To make the the decision about whether to skip the step, `GradScaler.step` [combines](https://github.com/pytorch/pytorch/blob/master/torch/cuda/amp/grad_scaler.py#L283) the `found_inf` tensors from all devices.

`_check_inf_per_device` does NOT automatically combine the `found_inf` tensors from all devices.  Your current code passes each device's `found_inf` tensor to the `stochastic_rounding_adam_step` kernel separately, which is incorrect.  If infs are found for params on one device but not others, the stochastic rounding kernel only skips the math for params on the device where infs were found.

Your logic should combine the `found_inf` tensors from all devices, then wrap the combined found inf tensor in `_MultiDeviceReplicator` for later use by the kernel.

A similar fix will be required in the other `SR*` optimizers.  In fact this requirement+logic are generic across optimizers.  You could make a helper function to wrap the logic above (that returns a replicator that wraps a combined found inf tensor), then all optimizers could call the helper function.  Here's an example that remains sync-free in the common case (logic is similar to [GradScaler.update](https://github.com/pytorch/pytorch/blob/1039b95ff056ba0e0118d40267c68a9df4535f7b/torch/cuda/amp/grad_scaler.py#L325-L334)):
```python
def _combined_found_inf_helper(optimizer, scaler):
    found_inf_dict = grad_scaler._check_inf_per_device(optimizer)
    # Combines found_inf tensors from all devices.  As in GradScaler.update(),
    # tensors are combined on the scale's device, which is an arbitrary but 
    # reasonable choice that avoids new context creation.
    found_infs = [f.to(scaler._scale.device, non_blocking=True) for f in found_inf_dict.values()]
    assert len(found_infs) > 0, "No inf checks were recorded in _check_inf_per_device."
    found_inf_combined = found_infs[0]
    if len(found_infs) > 1:
        for i in range(1, len(found_infs)):
            found_inf_combined += found_infs[i]
    return _MultiDeviceReplicator(found_inf_combined)
```
then within an optimizer e.g.
```python
found_inf = _combined_found_inf_helper(self, grad_scaler)
...
# same kernel call, but now it's safe, because found_inf accounts for infs from all devices
torch.stochastic_rounding_adam_step(..., found_inf.get(param.device), ...) 
```

crcrpar replied:
Thank you for the detailed and easy to understand explanation and suggestion!!!

ptrblck commented on torch/optim/sradamw.py:
```diff
@@ -0,0 +1,96 @@
+
+        if grad_scaler is not None:
+            found_inf = grad_scaler._check_inf_per_device(
+                self)[torch.device(torch.cuda.current_device())]
```
Same as above regarding `current_device`

ptrblck commented on torch/optim/srsgd.py:
```diff
@@ -0,0 +1,72 @@
+
+        if grad_scaler is not None:
+            found_inf = grad_scaler._check_inf_per_device(
+                self)[torch.device(torch.cuda.current_device())]
```
Same as above regarding `current_device`

crcrpar committed:
make SRAdam(W) compatible with Adam(W)

crcrpar committed:
cosmetic

crcrpar committed:
Add comment about test condition

zasdfgbnm commented on aten/src/ATen/native/cuda/stochastic_rounding.cuh:
```diff
@@ -0,0 +1,63 @@
+#ifndef _STOCHASTIC_ROUNDING_CUH_
```
nit:
```
#pragma once
```

zasdfgbnm commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,53 @@
+namespace native {
+
+template <typename input_t, typename output_t>
+__global__ void stochastic_rounding_kernel(
```
This would be a good use case of `distribution_unary_kernel` :)

zasdfgbnm commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,53 @@
+
+Tensor stochastic_rounding_cuda(const Tensor& input, Generator gen_) {
+
+  Tensor output = at::empty_like(input, input.options().dtype(kHalf), input.suggest_memory_format());
```
Do you need to check if `input` is contiguous? Because you are looping  on 0 to numel, and if the input has some broadcasting, then you will get out of boundary in your kernel.

crcrpar replied:
Yes, I do (I created this function mainly for the sake of unittest and I was not careful). thank you.

crcrpar replied:
Now checking the contiguity https://github.com/csarofeen/pytorch/pull/17/files#diff-feb2a89ccaa81c81b69bf4c91e8f60a3R26

zasdfgbnm commented on aten/src/ATen/native/cuda/StochasticRoundingAdam.cu:
```diff
@@ -0,0 +1,123 @@
+  TORCH_CHECK(max_exp_avg_sq.is_contiguous());
+
+  const int64_t numel = param.numel();
+  const int block_size = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
+  dim3 dim_block(block_size);
+  dim3 grid((numel + block_size - 1) / block_size);
+  grid.x = std::min((unsigned int)at::cuda::getCurrentDeviceProperties()->multiProcessorCount * blocks_per_sm, grid.x);
+
+  auto gen = get_generator_or_default<CUDAGeneratorImpl>(gen_, cuda::detail::getDefaultCUDAGenerator());
+
+  uint64_t counter_offset = ((numel + dim_block.x * grid.x - 1) / (block_size * grid.x)) * 4;
+  std::pair<uint64_t, uint64_t> rng_engine_inputs;
+  {
+    std::lock_guard<std::mutex> lock(gen->mutex_);
+    rng_engine_inputs = gen->philox_engine_inputs(counter_offset);
+  }
```
I see copy-pasted code appearing 3 times, is it possible to use the`calc_execution_policy` in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/DistributionTemplates.h#L37?

crcrpar replied:
Yeah, I don't think this copy-paste cool, however, it's impossible as of now because `calc_execution_policy`'s `offset` differs from ones in this PR.

https://github.com/pytorch/pytorch/blob/63e5058c885309f577a3871d18603eb5580b7996/aten/src/ATen/native/cuda/DistributionTemplates.h#L48-L49

zasdfgbnm replied:
OK

zasdfgbnm commented on test/test_stochastic_rounding.py:
```diff
@@ -0,0 +1,28 @@
+import math
```
If you add a new test file, you should add it to https://github.com/pytorch/pytorch/blob/master/test/run_test.py

crcrpar replied:
Thank you!

zasdfgbnm commented on torch/optim/sradam.py:
```diff
@@ -0,0 +1,115 @@
+            algorithm from the paper `On the Convergence of Adam and Beyond`_
+            (default: False)
+
+    .. _Adam\: A Method for Stochastic Optimization:
```
Where does the idea of stochastic rounding come from? Is there a paper to cite?

crcrpar replied:
Put a link in `torch.stochastic_rounding` docstring and now optimizers refer to it.
https://github.com/csarofeen/pytorch/pull/17/files#diff-cfb065e43691d37ba30f3856fe4352b3R7536

crcrpar committed:
use pragma once

crcrpar committed:
register stochastic rounding test

crcrpar committed:
return input when dtype is Half

crcrpar committed:
Update documentation

- add docstring to `torch.stochastic_rounding`
- refer `torch.stochastic_rounding` from stochastic rounding optimizers
  for details

crcrpar committed:
handle non-current devices

crcrpar committed:
Generator -> c10::optional<Generator>

mcarilli commented on torch/optim/__init__.py:
```diff
@@ -13,6 +13,9 @@
 from .sgd import SGD
+from .sradam import SRAdam
+from .sradamw import SRAdamW
+from .srsgd import SRSGD
```
It's unclear whether we want to expose stochastic rounding as separate optimizer classes or as constructor options to existing optimizers.  I don't think we need to make this choice now, though.  I think Facebook people should look at what you've got then we can decide together.

crcrpar committed:
combine `found_inf` tensors from all devices

mcarilli commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,62 @@
+
+  const int block = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block;
+  unsigned int grid = (numel + block - 1) / block;
```
This defeats the purpose of the grid-stride loop.  If you want to maximize the benefit of the grid stride loop, hardcode
`unsigned int grid = blocks_per_sm*(at::cuda::getCurrentDeviceProperties()->multiProcessorCount)`.

mcarilli commented on aten/src/ATen/native/cuda/StochasticRoundingAdam.cu:
```diff
@@ -0,0 +1,123 @@
+  const int block_size = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
+  dim3 dim_block(block_size);
+  dim3 grid((numel + block_size - 1) / block_size);
```
As for [stochastic_rounding_cuda](https://github.com/csarofeen/pytorch/pull/17/files#r420368297), this removes the benefit of the grid stride loop.  Hardcode
`dim3 grid(blocks_per_sm*(at::cuda::getCurrentDeviceProperties()->multiProcessorCount));`

mcarilli commented on aten/src/ATen/native/cuda/StochasticRoundingSGD.cu:
```diff
@@ -0,0 +1,93 @@
+  const int block_size = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
+  dim3 dim_block(block_size);
+  dim3 grid((numel + block_size - 1) / block_size);
```
As for [stochastic_rounding_cuda](https://github.com/csarofeen/pytorch/pull/17/files#r420368297), this removes the benefit of the grid stride loop.  Hardcode
`dim3 grid(blocks_per_sm*(at::cuda::getCurrentDeviceProperties()->multiProcessorCount));`

mcarilli commented on aten/src/ATen/native/cuda/StochasticRoundingSGD.cu:
```diff
@@ -0,0 +1,93 @@
+    float weight = static_cast<float>(weights[i]);
+    float gradient = static_cast<float>(gradients[i]) * (*inv_scale);
+    float velocity = static_cast<float>(momentum_buffer[i]);
+    float4 random_values = curand_uniform4(&state);
```
you generate 4 rng and only use 2.  I don't think that's a big problem though.

mcarilli commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  const int block = 256;
+  const int blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block;
```
This is only correct if the kernel's number of registers per thread is <= 32, otherwise register pressure limits your occupancy.  You can recompile kernels with `-ptxas-options=-v` as an nvcc option and nvcc will print how many registers they use (this is easiest to do with the kernels in an extension, I'm not sure how you would pass that option to nvcc in a pytorch build).

mcarilli commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,62 @@
+
+  AT_DISPATCH_FLOATING_TYPES(
+    input.scalar_type(),  "stochastic_rounding_cuda", [&] {
+      stochastic_rounding_kernel<scalar_t, at::Half><<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(
```
My biggest concern is, upstream will probably ask you to rewrite this with TensorIterator in some form, as @zasdfgbnm hinted.

mcarilli commented on aten/src/ATen/native/cuda/stochastic_rounding.cuh:
```diff
@@ -0,0 +1,60 @@
+
+// Natalia magic
+template <typename scalar_t>
+__device__ __forceinline__ scalar_t round_stochastically(float x, float random_value) {
```
Right now, this function only supports rounding to fp16.  I think its syntax is misleading.

To make the usage clearer, and to establish an API that supports stochastic rounding to other types in the future, I think you should define it as follows:
```
template<typename round_to_prec, typename out_type, typename in_type=float>
struct round_stochastically {
  static_assert(false, "round_stochastically only supports round_to_prec=at::Half");
  __device__ __forceinline__ out_type operator()(in_type x, float random_value) {}
};

template <typename out_type, typename in_type=float>
struct round_stochastically<at::Half, out_type, in_type> {
  __device__ __forceinline__ at::Half operator()(in_type x, float random_value) {
    // what we have now
  }
}
```

Then the caller should say
```
weights[i] = round_stochastically<at::Half, scalar_t>(weight, random_values.x);
```

crcrpar replied:
L.59' s `maybe_upcast` does cast from `at::Half` to `float`/`double` if necessary and stochastic rounding SGD & Adam kernel use this functionality.

mcarilli replied:
We might want to support stochastic rounding to other formats (like bfloat16) later.  The API should allow the caller to set a type that determines the rounding precision, even if the actual rounding code for that precision isn't implemented yet.

crcrpar replied:
Understand. Updated.

mcarilli commented on aten/src/ATen/native/cuda/StochasticRounding.cu:
```diff
@@ -0,0 +1,62 @@
+  curand_init(seed_and_offset.first, tid, seed_and_offset.second, &state);
+
+  for (int64_t i = tid; i < numel; i += blockDim.x * gridDim.x) {
+    float inp = static_cast<float>(input[i]);
```
If you make these changes https://github.com/csarofeen/pytorch/pull/17/files#r420422970 the cast to float won't be necessary.

crcrpar committed:
Make round_stochastically flexible for future another dtype support

mcarilli commented on aten/src/ATen/native/cuda/stochastic_rounding.cuh:
```diff
@@ -44,17 +44,24 @@ __device__ __forceinline__ float get_delta_fp16(float x) {
+    if (x == 0.0) {
+      return out_type(0.0);
+    }
+    float delta = get_delta_fp16(static_cast<float>(x));
```
Hardcoding float here is probably fine IMO, but natalia may ask you to change this to `in_type` (which might require making `get_delta_fp16` a template), and replace the `__float2half_rz` call with a wrapper function that has several overloads and the float overload calls `__float2half_rz`.
